crates/bevy_render/macros/src/as_bind_group.rs

```diff
@@ -385,5 +393,451 @@ pub fn derive_as_bind_group(input: TokenStream) -> TokenStream {
+}
+
+impl Parse for BindingMeta {
+    fn parse(input: ParseStream) -> Result<Self> {
+        if input.peek2(Token![,]) {
+            input.parse().map(Self::IndexWithOptions)
+        } else {
+            input.parse().map(Self::IndexOnly)
+        }
+    }
+}
+
+impl Parse for BindingIndexOptions {
+    fn parse(input: ParseStream) -> Result<Self> {
+        Ok(Self {
+            lit_int: input.parse()?,
+            _comma: input.parse()?,
+            meta_list: input.parse_terminated(NestedMeta::parse)?,
+        })
+    }
+}
+
+impl Parse for UniformBindingMeta {
+    fn parse(input: ParseStream) -> Result<Self> {
+        Ok(Self {
+            lit_int: input.parse()?,
+            _comma: input.parse()?,
+            ident: input.parse()?,
+        })
+    }
+}
+
+fn get_uniform_binding_attr(attr: &syn::Attribute) -> Result<(u32, Ident)> {
+    let uniform_binding_meta = attr.parse_args_with(UniformBindingMeta::parse)?;
+
+    let binding_index = uniform_binding_meta.lit_int.base10_parse()?;
+    let ident = uniform_binding_meta.ident;
+
+    Ok((binding_index, ident))
+}
+
+fn get_binding_nested_attr(attr: &syn::Attribute) -> Result<(u32, Vec<NestedMeta>)> {
+    let binding_meta = attr.parse_args_with(BindingMeta::parse)?;
+
+    match binding_meta {
+        BindingMeta::IndexOnly(lit_int) => Ok((lit_int.base10_parse()?, Vec::new())),
+        BindingMeta::IndexWithOptions(BindingIndexOptions {
+            lit_int,
+            _comma: _,
+            meta_list,
+        }) => Ok((lit_int.base10_parse()?, meta_list.into_iter().collect())),
+    }
+}
+
+#[derive(Default)]
+enum ShaderStageVisibility {
+    #[default]
+    All,
+    None,
+    Flags(VisibilityFlags),
+}
+
+#[derive(Default)]
+struct VisibilityFlags {
+    vertex: bool,
+    fragment: bool,
+    compute: bool,
+}
+
+impl ToTokens for ShaderStageVisibility {
+    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
+        tokens.extend(match self {
+            ShaderStageVisibility::All => quote! { ShaderStages::all() },
+            ShaderStageVisibility::None => quote! { ShaderStages::NONE },
+            ShaderStageVisibility::Flags(flags) => {
+                let mut quoted = Vec::new();
+
+                if flags.vertex {
+                    quoted.push(quote! { ShaderStages::VERTEX });
+                }
+                if flags.fragment {
+                    quoted.push(quote! { ShaderStages::FRAGMENT });
+                }
+                if flags.compute {
+                    quoted.push(quote! { ShaderStages::COMPUTE });
+                }
+
+                quote! { #(#quoted)|* }
+            }
+        });
+    }
+}
+
+const VISIBILITY: Symbol = Symbol("visibility");
+const VISIBILITY_VERTEX: Symbol = Symbol("VERTEX");
+const VISIBILITY_FRAGMENT: Symbol = Symbol("FRAGMENT");
+const VISIBILITY_COMPUTE: Symbol = Symbol("COMPUTE");
+const VISIBILITY_ALL: Symbol = Symbol("ALL");
+const VISIBILITY_NONE: Symbol = Symbol("NONE");
```

Comment:
I think they're capitalised in the code more because they are constants. So just a coding convention. I agree that lowercase would feel more consistent here.